File: src/main/java/january_2025/MyPow.java
package january_2025;

import java.util.HashMap;
import java.util.Map;

public class MyPow {

    public static void main(String[] args) {
        MyPow myPow = new MyPow();
        double result = myPow.myPow(2.0, 5);
        System.out.println(result);

//        double result2 = myPow.myPow(2.0, -200000000);
//        System.out.println(result2);
    }

    public double myPow(double x, int n) {
        if (n == 0) {
            return 1.0;
        }
        if (n == 1) {
            return x;
        }
        double res = helper(x, Math.abs((long) n));
        return n < 0 ? 1 / res : res;
    }

    private double helper(double x, long n) {
        if (n == 0) {
            return 1;
        }
        double half = helper(x, n / 2);
        return n % 2 == 0 ? half * half : x * half * half;
    }



}

File: python/my_pow.py
class Solution(object):
    # O(log(n)) time | O(log(n)) space
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        if n == 0:
            return 1
        
        def helper(a, b):
            if b == 1:
                return a
            half = helper(a, b // 2)
            return half * half if b % 2 == 0 else a * half * half

        return helper(x, n) if n > 0 else 1 / helper(x, -1 * n)

solution = Solution()
print(solution.myPow(2.0, 10))